src/lib.rs

```diff
@@ -50,46 +50,55 @@ pub fn derive_deref_mut(input: TokenStream) -> TokenStream {
 fn parse_fields(item: &syn::DeriveInput, mutable: bool) -> (syn::Type, proc_macro2::TokenStream) {
     let trait_name = if mutable { "DerefMut" } else { "Deref" };
     let fields = match item.data {
-        syn::Data::Struct(ref body) => body.fields.iter().collect::<Vec<&syn::Field>>(),
+        syn::Data::Struct(ref body) =>
+            body.fields.iter()
+                .filter(|field| {
+                    if let Type::Path(TypePath { path: Path { segments, .. }, .. }) = &field.ty {
+                        if segments
+                            .last()
+                            .expect("Expected path to have at least one segment")
+                            .ident == "PhantomData"
+                        {
+                            false
+                        } else {
+                            true
+                        }
+                    } else {
+                        true
+                    }
+                })
+                .collect::<Vec<&syn::Field>>(),
         _ => panic!("#[derive({})] can only be used on structs", trait_name),
     };
 
-    let field_ty = match fields.len() {
-        1 => Some(fields[0].ty.clone()),
-        2 => {
-            if let Type::Path(TypePath { path: Path { segments, .. }, .. }) = &fields[1].ty {
-                if segments
-                    .last()
-                    .expect("Expected path to have at least one segment")
-                    .value()
-                    .ident == "PhantomData"
-                {
-                    Some(fields[0].ty.clone())
-                } else {
-                    None
+    let target_field = match fields.len() {
+        1 => fields[0],
+        _ => {
+            // Look for attribute
+            let mut targets = Vec::new();
+            for field in fields.iter() {
+                if let Some(_) = field.attrs.iter().find(|attr| attr.path.is_ident("deref_target")) {
+                    targets.push(field);
                 }
-            } else {
-                None
             }
-        },
-        _ => None,
+            match targets.len() {
+                0 => panic!("#[derive({})]: since there is more than one field, #[deref_target] is needed on a field", trait_name),
+                1 => targets[0],
+                _ => panic!("#[derive({})]: only one #[deref_target] allowed", trait_name),
+            }
+
+        }
```

Comment:
It seems like we can restructure this using `Option::and_then` and `Vec::get`? I don't think the behavior when there's more than one of these attributes is worth the complexity

Reply:
just `*targets.get(0).expect("#[deref_target] expected on one of the fields")`?

Reply:
(changed it)

Reply:
I don't agree: this looks a lot like a slient fail to me, as as there should not be two such flags on a single structure.
```rust
let mut targets = fields.iter().filter(|field| field.attrs.iter().any(|attr| attr.path.is_ident("deref_target")));
match (targets.next(), targets.next()) {
    (Some(target), None) => target,
    (None, _) => panic!("#[derive({})]: since there is more than one field, #[deref_target] is needed on a field", trait_name),
    (Some(_), Some(_)) => panic!("#[derive({})]: only one #[deref_target] allowed", trait_name),
}
```
is a much simpler expression of that though.